### PR TITLE
GAWB-850: FireCloud Metrics Admin Endpoint

### PR DESCRIPTION
**Pull request by MatthewBemis**
PR notes in progress
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Make sure liquibase is updated if appropriate
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] **Submitter**: If PR includes new or changed db queries, include the explain plans in the description
- [x] Tell  that the PR exists if he wants to look at it **(apply requires_doge label)**
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x]  sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [ ] **Submitter**: Squash commits, rebase if necessary
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Merge to develop 
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: Check configuration files in Jenkins in case they need changes
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by MatthewBemis** [comment]
Going to be sending these metrics out to some folks from CGA for review so some queries may change but for now here are the explain plans. Yes, there are a lot of full table scans because we don't have indexes on the fields that these queries rely on. In dev, this endpoint is snappy, but in the future it may not be. Should we add indexes or let it slide until it becomes a problem? CGA will be calling this endpoint only once per week.

`select min(count), max(count), avg(count), stddev(count)
                from (select count(1) as count from SUBMISSION
                where DATE_SUBMITTED between 2010-10-10 and 2016-10-10
                group by SUBMITTER) as counts;`
screen shot 2016-07-22 at 1 57 47 pm

## 

`select min(seconds), max(seconds), avg(seconds), stddev(seconds)
                from (
                  select TIMESTAMPDIFF(SECOND, s.DATE_SUBMITTED, a.timestamp) as seconds
                    from SUBMISSION s join AUDIT_SUBMISSION_STATUS a on s.ID=a.submission_id
                    where s.STATUS in ("Done") and a.STATUS in ("Done")
                    and a.timestamp between 2010-10-10 and 2016-10-10
                ) as runtimes;`
screen shot 2016-07-22 at 1 59 06 pm

## 

`select count(1) from SUBMISSION
                where DATE_SUBMITTED between 2010-10-10 and 2016-10-10;`
screen shot 2016-07-22 at 2 00 22 pm

## 

`select count(distinct SUBMITTER) from SUBMISSION
                where DATE_SUBMITTED between 2010-10-10 and 2016-10-10;`
screen shot 2016-07-22 at 2 01 45 pm

## 

`select min(count), max(count), avg(count), stddev(count)
                from (
                  select count(1) as count from WORKFLOW w
                    join SUBMISSION s on s.ID=w.SUBMISSION_ID
                    where s.DATE_SUBMITTED between 2010-10-10 and 2016-10-10
                    group by s.SUBMITTER
                ) as counts;`
screen shot 2016-07-22 at 2 03 14 pm

## 

`select min(count), max(count), avg(count), stddev(count)
                from (
                  select count(1) as count from WORKFLOW w
                    join SUBMISSION s on s.ID=w.SUBMISSION_ID
                    where s.DATE_SUBMITTED between 2010-10-10 and 2016-10-10
                    group by w.SUBMISSION_ID
                ) as counts;`
screen shot 2016-07-22 at 2 05 57 pm

## 

`select min(seconds), max(seconds), avg(seconds), stddev(seconds)
                from (
                  select TIMESTAMPDIFF(SECOND, MIN(a.timestamp), MAX(a.timestamp)) as seconds
                    from WORKFLOW w join AUDIT_WORKFLOW_STATUS a on w.ID=a.workflow_id
                    where w.STATUS in ("Succeeded") and a.STATUS in ("Queued","Succeeded")
                    and a.timestamp between 2010-10-10 and 2016-10-10 group by a.workflow_id
                ) as runtimes;`
screen shot 2016-07-22 at 2 06 58 pm

## 

`select count(1) from WORKFLOW w
                join SUBMISSION s on s.ID=w.SUBMISSION_ID
                where s.DATE_SUBMITTED between "2010-10-10" and "2016-10-10";`
screen shot 2016-07-22 at 2 08 50 pm

## 

`select count(1) from USER;`
screen shot 2016-07-22 at 2 09 22 pm

**Comment by jmthibault79** [comment]
I am OK with not optimizing this endpoint's performance.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala [hunk 1]
```diff
@@ -569,6 +570,47 @@ trait WorkflowComponent {
...
+                    group by w.SUBMISSION_ID
+                ) as counts""".as[SummaryStatistics].head
+      }
+
+      def workflowRunTimeQuery(startDate: String, endDate: String) = {
+        sql"""select min(seconds), max(seconds), avg(seconds), stddev(seconds)
+                from (
+                  select TIMESTAMPDIFF(SECOND, MIN(a.timestamp), MAX(a.timestamp)) as seconds
+                    from WORKFLOW w join AUDIT_WORKFLOW_STATUS a on w.ID=a.workflow_id
+                    where w.STATUS in ("Succeeded") and a.STATUS in ("Queued","Succeeded")
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala:603]: Is there a way to guarantee that the audit table returns both a Queued and Succeeded row?  If there's only one, it will add spurious zeroes to the stats.  Maybe that's how to do it: exclude diff=zero

File: src/main/scala/org/broadinstitute/dsde/rawls/model/StatisticsModel.scala [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+object StatisticsJsonSupport extends JsonSupport {
+  implicit val SingleStatisticFormat = jsonFormat1(SingleStatistic)
+  implicit val SummaryStatisticsFormat = jsonFormat4(SummaryStatistics)
+
+  implicit object StatisticFormat extends RootJsonFormat[Statistic] {
+    override def write(obj: Statistic): JsValue = obj match {
+      case SingleStatistic(value) => JsObject(Map("value" -> JsNumber(value)))
+      case SummaryStatistics(min, max, mean, stddev) => JsObject(Map("min" -> JsNumber(min), "max" -> JsNumber(max), "mean" -> JsNumber(mean), "stddev" -> JsNumber(stddev)))
+    }
+    override def read(json: JsValue): Statistic = json match {
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/model/StatisticsModel.scala]: Why is this needed?

File: src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponentSpec.scala [hunk 3]
```diff
@@ -131,6 +131,18 @@ class SubmissionComponentSpec extends TestDriverComponentWithFlatSpecAndMatchers
     }
   }
 
+  //if this unit test breaks, chances are you have added a submission to the test data which has changed the values below
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/SubmissionComponentSpec.scala:134]: I don't like brittle tests.  We're going to be updating this one a lot as the test data changes, and that won't tell us much about whether these queries are working.

Could we use a different known data set?  Maybe a TestData which won't change, specifically for these.

File: src/main/scala/org/broadinstitute/dsde/rawls/model/StatisticsModel.scala (hunk 2 above)
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/rawls/model/StatisticsModel.scala]: Compiler was angry that I wasn't overriding the read method. A read is never actually needed so I threw in a useless implementation of it.

Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/model/StatisticsModel.scala]: In that case, can you make it even more useless?  Remove the match.  Maybe even make it explicitly unimplemented with the special `???` operator, like `override def read(json: JsValue) = ???` if the compiler will allow it.

Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/model/StatisticsModel.scala]: Or could this extend `RootJsonWriter` instead of `RootJsonFormat` ?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala (hunk 1 above)
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala:603]: assuming our DB triggers stay intact, having a queued and succeeded row is guaranteed. the queued status should always be guaranteed because that's entered when the workflow is first created, and the succeeded status will always be guaranteed because we're only looking at workflows that have the succeeded status to begin with

if you feel strongly about excluding zeroes just in case, i'll add it but i don't think it'll be an actual problem

Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkflowComponent.scala:603]: ok works for me as-is

**Comment by MatthewBemis** [comment]
@jmthibault79 back to you
- added ConstantTestData for use in these statistics tests
- implemented the json read as ???

**Comment by jmthibault79** [comment]
👍

**Comment by MatthewBemis** [comment]
Sign-offs acknowledged. This is also awaiting a quick review from someone in CGA.
